htlcswitch/switch.go

```diff
@@ -1075,6 +1079,23 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
 			return s.failAddPacket(packet, linkErr.WireMessage(), addErr)
 		}
 
+		// We have found an outgoing link with a forwarding policy that
+		// the htlc satisfies. Before we proceed with forwarding, check
+		// whether the htlc will produce a circular forward through the
+		// same channel, and fail it if the switch does not allow
+		// forwards of this nature.
+		if linkErr := s.CheckCircularForward(
```

Comment:
I am wondering how this check works together with the previous non-strict forwarding decision. Could it be that the same in and out chan was requested, but that the out channel for some reason (concurrent payment perhaps?) doesn't have enough balance and another channel to the same node is selected?

Reply:
I think that it's fair to treat these payments as 'suspicious' because they are taking an inefficient route. While there may be some custom use cases out there, node operators who are expecting to use them should configure their node to allow circular routes. 

If somebody tries to route a payment over the same channel (and we do not allow that), we could just catch that early and not apply non-strict forwarding? Which would just mean moving the check up in `handlePacketForward` and failing early.

Reply:
Yes moving it up sounds good. Or maybe basing the check on a pubkey comparison, maybe that is even better.

Reply:
Disallowing based on pubkey would mean that people can't do 2 hop circular rebalances, which I think would cause some issues for existing nodes if their lnd peers suddenly stopped letting them rebalance.

Reply:
Two hop circular rebalances are problematic anyway, because of non-strict forwarding. You may just as well get back your money through the same channel.

Comparing on the exact chan ids is also a risk for peers that have multiple channels between them. The ping pong route can still be executed. I think it is more important to prevent that than to allow the two hop rebalance which isn't reliable anyway.

Reply:
It is unreliable but I think a lot of people rely on it at the moment. Would say that we'd probably disable by default if we disallow to the same peer, so that we don't break a lot of people's setups. 

I think that single channel is still a valuable defense to have. Without it, an attacker can lock up funds in every channel on your node by zig zagging back and fourth through each channel. The takes `capacity/(max htlc size*19)` payments to lock up your balance (assuming one hop where they lock up their own balance to route the first hop to your node. I think that to one peer would need wider discussion, @alexbosworth has an operational requirement for that ability, for example.

Reply:
Yes, single channel is better than nothing. I have no data myself on how many people rely on two hop rebalancing. Detecting peer-level ping-pong can be implemented later if there are doubts.

One way to lock up the channel is to use up all its capacity, but another way is to use up all commitment tx slots. There is a max number of htlcs that the tx can hold. That way it costs only negligible money to do the attack.